drivers/regulator/regulator_cp9314.c

```diff
@@ -0,0 +1,524 @@
+	uint8_t reg_val;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, CP9314_REG_CTRL4, &reg_val);
+	if (ret < 0) {
+		return ret;
+	}
+
+	*mode = reg_val & CP9314_MODE_MASK;
+
+	return 0;
+}
+
+static int regulator_cp9314_disable(const struct device *dev)
+{
+	const struct regulator_cp9314_config *config = dev->config;
+
+	if (config->en_pin.port != NULL) {
+		return gpio_pin_set_dt(&config->en_pin, 0);
+	}
+
+	return i2c_reg_update_byte_dt(&config->i2c, CP9314_REG_CTRL1, CP9314_CP_EN, 0);
+}
+
+static int regulator_cp9314_enable(const struct device *dev)
+{
+	const struct regulator_cp9314_config *config = dev->config;
+	uint8_t value;
+	int ret;
+
+	ret = i2c_reg_read_byte_dt(&config->i2c, CP9314_REG_CONVERTER, &value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (value & CP9314_ACTIVE_STS) {
+		return 0;
+	}
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, CP9314_REG_LION_INT_MASK_2, CP9314_CLEAR_INT,
+				     CP9314_CLEAR_INT);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, CP9314_REG_LION_INT_MASK_2, CP9314_CLEAR_INT, 0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	if (config->en_pin.port != NULL) {
+		return gpio_pin_set_dt(&config->en_pin, 1);
+	}
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, CP9314_REG_CTRL1, CP9314_CP_EN, CP9314_CP_EN);
+	if (ret < 0) {
+		LOG_ERR("Unable to set CP_EN");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int cp9314_cfg_sync(const struct device *dev)
+{
+	const struct regulator_cp9314_config *config = dev->config;
+	struct regulator_cp9314_data *data = dev->data;
+	uint8_t value = 0;
+	int ret;
+
+	if (data->sync_role == CP9314_ROLE_HOST) {
+		value = CP9314_SYNC_HOST_EN;
+	}
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, CP9314_REG_CTRL4, CP9314_SYNC_HOST_EN, value);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, CP9314_REG_CTRL4, CP9314_SYNC_FUNCTION_EN,
+				     CP9314_SYNC_FUNCTION_EN);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return i2c_reg_update_byte_dt(&config->i2c, CP9314_REG_CTRL4, CP9314_FRC_SYNC_MODE,
+				      CP9314_FRC_SYNC_MODE);
+}
+
+static int regulator_cp9314_b0_init(const struct device *dev)
+{
+	const struct regulator_cp9314_config *config = dev->config;
+	int ret;
+
+	for (size_t i = 0U; i < ARRAY_SIZE(b0_reg_patch); i++) {
+		ret = i2c_reg_update_byte_dt(&config->i2c, b0_reg_patch[i].reg_addr,
+					     b0_reg_patch[i].mask, b0_reg_patch[i].value);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+
+	return 0;
+}
+
+static int cp9314_do_soft_reset(const struct device *dev)
+{
+	const struct regulator_cp9314_config *config = dev->config;
+	int ret;
+
+	ret = i2c_reg_write_byte_dt(&config->i2c, CP9314_REG_CRUS_CTRL, CP9314_CRUS_KEY_SOFT_RESET);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = i2c_reg_update_byte_dt(&config->i2c, CP9314_REG_TEST_MODE_CTRL, CP9314_SOFT_RESET_REQ,
+				     CP9314_SOFT_RESET_REQ);
+	if (ret < 0) {
+		return ret;
+	}
+
+	k_msleep(CP9314_SOFT_RESET_DELAY_MSEC);
```

Comment:
Please add an ifdef here so it works with multithreading disabled, e.g:
https://github.com/search?q=repo%3Azephyrproject-rtos%2Fzephyr+k_sleep+path%3A%2F%5Edrivers%5C%2Fregulator%5C%2F%2F&type=code

Reply:
understood, thanks for the tip.